rust/connlib/tunnel/src/gateway.rs

```diff
@@ -197,6 +209,16 @@ impl GatewayState {
             return None;
         };
 
+        if let Err(e) = peer.ensure_allowed_source(&packet) {
+            // Note: this can happen with apps such as cURL that if started before the tunnel routes are address
+            // source ips can be sticky.
+            tracing::warn!(%conn_id, %local, %from, "Invalid packet source: {e}");
+
+            return None;
+        }
+
+        let packet = peer.translate_proxy_to_real(packet, now)?;
+
         if let Err(e) = peer.ensure_allowed(&packet) {
```

Comment:
We are calling 3 functions on `peer` here. Should we introduce a `peer.decapsulate`? (or `translate` / `transform`?)